### ISSUE TITLE
Prepared Statements for SQL destination

### ISSUE THREAD

**Issue by Wernfried** [balabit/syslog-ng#1018]
A great enhancement would be if syslog-ng can use prepared statements for SQL inserts.

My tests with SQLite shows that prepared statements are around twice as fast as static statements. Various web pages are showing the same figure. On Oracle prepared statements also have a big performance impact. For MySQL and PostgreSQL I do not know the impact.

An application which claims to be "high-performance" or "real-time" should support such feature. I am not familiar with C/C++, perhaps at least for SQLite an implementation would be easier when you use the SQLite C/C++ Interface: https://www.sqlite.org/cintro.html

Maybe other database provide similar API's.

**Comment by bazsi** [comment]
The idea is sound, but I dont know if libdbi supports prepared statements
or not. Libdbi is holding us back in a few ways though.

**Comment by Wernfried** [comment]
I made a small benchmark comparison with Perl, result is following:

```
SQLite  Prepared Statements     No Transaction          4.305s -> 23 rows/s
SQLite  Static Statements       No Transaction          4.914s -> 20 rows/s
SQLite  Prepared Statements     With Transaction        0.951s -> 105138 rows/s <- !!!
SQLite  Static Statements       With Transaction        1.779s -> 56200 rows/s

Oracle  Prepared Statements     No Transaction          1.095s -> 913 rows/s
Oracle  Static Statements       No Transaction          1.200s -> 833 rows/s
Oracle  Prepared Statements     With Transaction        0.362s -> 2761 rows/s
Oracle  Static Statements       With Transaction        0.480s -> 2082 rows/s

```
You see the most important feature is to use transactions (which is the case for syslog-ng - unlike  competing rsyslog). For SQLite you gain about 100%. For Oracle it makes less difference, however you have a significant improvement in performance.

Looks like libdbi does not support prepared statements (at least not for SQLite), see this code snippet in file dbd_sqlite3.c:

```
dbi_result_t *dbd_query_null(dbi_conn_t *conn, const unsigned char *statement, size_t st_length) {
  /* todo: implement using sqlite3_prepare and friends */
  return NULL;
} 

```

I found this [SOCI - The C++ Database Access Library](http://soci.sourceforge.net/) , no idea if this could be an alternative.


Here is the Perls script I used for performance comparison:

```
use DBI;
use Time::HiRes qw( time );

	my ($dbh, $sth, $inserts, $sttime, $elapse);

	$dbh = DBI->connect("dbi:SQLite:dbname=$ARGV[0]", undef, undef, { RaiseError => 1} );

	$inserts = 100;
	$dbh->do("delete from messages");
	$sttime = time();
	$sth = $dbh->prepare("insert into messages values (CURRENT_TIMESTAMP,?,?,?,?,?,?,?,?,?,?)");
	for ( my $i = 1; $i <= $inserts; $i++ ) {
		$sth->execute( ("syslog", "notice", "ALTS01TBMFW01", $i, "pba-create", "10.x.x.x", "212.x.x.x", 5245, 5308, undef) );
	}
	$elapse = time() - $sttime;
	printf("SQLite\tPrepared Statements\tNo Transaction  \t%.3fs -> %d rows/s\n", $elapse, $inserts / $elapse );

	$dbh->do("delete from messages");
	$sttime = time();
	for ( my $i = 1; $i <= $inserts; $i++ ) {
		$dbh->do(qq/insert into messages values (CURRENT_TIMESTAMP,"syslog", "notice", "ALTS01TBMFW01", $i, "pba-create", "10.x.x.x", "212.x.x.x", 5245, 5308,NULL)/);
	}
	$elapse = time() - $sttime;
	printf("SQLite\tStatic Statements  \tNo Transaction  \t%.3fs -> %d rows/s\n", $elapse, $inserts / $elapse );


	$inserts = 100000;
	$dbh->do("delete from messages");
	$sttime = time();
	$dbh->begin_work;
	$sth = $dbh->prepare("insert into messages values (CURRENT_TIMESTAMP,?,?,?,?,?,?,?,?,?,?)");
	for ( my $i = 1; $i <= $inserts; $i++ ) {
		$sth->execute( ("syslog", "notice", "ALTS01TBMFW01", $i, "pba-create", "10.x.x.x", "212.x.x.x", 5245, 5308, undef) );
	}
	$dbh->commit;	
	$elapse = time() - $sttime;
	printf("SQLite\tPrepared Statements\tWith Transaction\t%.3fs -> %d rows/s\n", $elapse, $inserts / $elapse );

	$dbh->do("delete from messages");
	$sttime = time();
	$dbh->begin_work;
	for ( my $i = 1; $i <= $inserts; $i++ ) {
		$dbh->do(qq/insert into messages values (CURRENT_TIMESTAMP,"syslog", "notice", "ALTS01TBMFW01", $i, "pba-create", "10.x.x.x", "212.x.x.x", 5245, 5308,NULL)/);
	}
	$dbh->commit;	
	$elapse = time() - $sttime;
	printf("SQLite\tStatic Statements  \tWith Transaction\t%.3fs -> %d rows/s\n", $elapse, $inserts / $elapse );

	$dbh->disconnect;  

```

**Comment by bazsi** [comment]
Hi,

This really seems convincing, however I can't really commit my time to
this. How difficult would it be to add the missing code to dbd_sqlite.c? I
guess the problem is that preparing the statements every time dbd_query()
is invoked would make no sense as the perf improvement comes from the fact
that the same statement is compiled only once, right?

If that's true, libdbi would have to add interfaces to prepare statements
(e.g. call to prepare a statement, and the ability to run a query with a
prepared statement), which it currently lacks completely.

These can be the potential solutions:
1) clone afsql, drop the generic libdbi calls and replace them with sqlite
specific calls and create an sqlite() specific destination
2) add prepared statements to libdbi (I wouldn't do this as libdbi is
basically dead)
3) evaluate and use an alternative SQL wrapper library (like SOCI)
4) use the python destination and interface SQL indirectly, this is
probably slow.

syslog-ng really uses a very narrow set of features of an SQL interface and
if we do 1) properly, introducing an interface layer that allows
libdbi-like swapping of sql implementations would not be impossible.

The problem is that I can't commit time to this right now. Can you?


-- 
Bazsi

**Comment by Wernfried** [comment]
Hi,

I have no clue about C/C++, otherwise I would consider to contribute.

But I agree with you, option 1) seems to be the most reasonable solution, option 3) also would to be reasonable in my opinion. - Just in case somebody likes to work on this topic.

Best Regards

**Comment by fekete-robert** [comment]
And what about implementing an SQL interfacte in Python? It seems that there are some solutions in Python that support prepared statements in a way (though I'm not sure about the performance gain of prepared statements vs. using Python instead of C), for example, for MySQL: http://stackoverflow.com/questions/27649759/using-prepared-statements-with-mysql-in-python
